Introduce Crystal::Hasher for computing a `hash` value

asterite opened the pull request:
This PR is a preparation to better implement the `hash` method for Crystal objects.

Instead of asking types to implement `hash`, we now provide a default implementation and ask subclasses to implement `hash(hasher)`. A hasher provides a safe way to compute a hash, and can also compute a different value for the same program at different runs to prevent DoS attacks. So this fixes #2817

The current implementation lives in `Crystal::Hasher` and it's not public. The idea is not to leak this information. For example in Ruby you can't pass a hasher to methods, and you can't choose a hasher. This is all decided by Ruby's runtime, and hidden, and this is good because it removes the burden from the user.

The implementation here is safe(r) because it randomizes the initial hash value. However, I'm sure we could use SipHash or FNV or some other algorithm. But this is just a start. Next PRs should just touch `Crystal::Hasher` (and probably `BigInt` and others because the hash is not correct there).

Fixes #2817
Fixes #4578

asterite commented:
This implements the code organization part of https://github.com/crystal-lang/crystal/pull/4675 but not the actual "good" implementation. But this PR is simpler to review.

asterite commented:
Also fixes https://github.com/crystal-lang/crystal/issues/4578

RX14 commented:
@asterite I don't think "fixes #foo" outside the main PR body actually gets github to close the issues (you can actually see on the issue page if a PR will close it now, which is cool)

asterite commented:
@RX14 Thanks, updated!

RX14 commented on src/crystal/hasher.cr:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def float(value)
+    @value = @value * 31 + value.to_f64.unsafe_as(UInt64)
```
I think we should at use [this](https://llvm.org/docs/LangRef.html#llvm-canonicalize-intrinsic) on the floats.

asterite replied:
We can leave it for a next PR

RX14 replied:
TODO then?

asterite replied:
There's a big todo at the top of the struct. On fact in this PR I could have used zero as a seed, but I wanted to see that it worked with it.

asterite replied:
I know what you are asking is not exactly covered by the above TODO. Still, I'd like to tackle these issues one at a time. After we merge this, I'll probably wait for someone (you? :-)) to use float normalization, though I'll require a test to see why it's needed. Then another PR can tackle another (better) hash algorithm, etc.

akzhan replied:
`canonicalize` is OK, but implementation proposed by @funny-falcon (PyHash inspired) is more robust and generic normalization that takes in care canonicalization too. Will propose it as separate PR.

straight-shoota commented on src/enum.cr:
```diff
@@ -274,9 +274,8 @@ struct Enum
-  # Returns a hash value. This is the hash of the underlying value.
-  def hash
-    value.hash
+  def hash(hasher)
```
Missing doc

asterite replied:
I'll add a "See `Object#hash(hasher)` to all of these

straight-shoota commented on src/float.cr:
```diff
@@ -86,6 +86,10 @@ struct Float
     end
   end
 
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/hash.cr:
````diff
@@ -706,18 +708,19 @@ class Hash(K, V)
-  # ```
-  def hash
-    hash = size
+  def hash(hasher)
````
Missing doc

straight-shoota commented on src/bool.cr:
```diff
@@ -42,8 +42,8 @@ struct Bool
   end
 
   # Returns a hash value for this boolean: 0 for `false`, 1 for `true`.
```
These fixed hash value should be removed.

straight-shoota commented on src/char.cr:
```diff
@@ -415,8 +415,8 @@ struct Char
   end
 
   # Returns this char's codepoint.
```
This doc string is no longer correct.

straight-shoota commented on src/class.cr:
```diff
@@ -3,8 +3,8 @@ class Class
 
-  def hash
-    crystal_type_id
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/indexable.cr:
```diff
@@ -271,13 +271,12 @@ module Indexable(T)
-  def hash
-    reduce(31 * size) do |memo, elem|
-      31 * memo + elem.hash
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/int.cr:
```diff
@@ -316,8 +316,8 @@ struct Int
 
-  def hash
-    self
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/named_tuple.cr:
```diff
@@ -159,16 +159,12 @@ struct NamedTuple
-  # See also: `Object#hash`.
-  def hash
-    hash = 31 * size
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/nil.cr:
```diff
@@ -68,8 +68,8 @@ struct Nil
   end
 
   # Returns `0`.
```
This value is no longer valid.

straight-shoota commented on src/proc.cr:
```diff
@@ -181,8 +181,8 @@ struct Proc
 
-  def hash
-    internal_representation.hash
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/reference.cr:
```diff
@@ -51,8 +51,8 @@ class Reference
   end
 
   # Returns this reference's `object_id` as the hash value.
```
The described value is not valid anymore.

straight-shoota commented on src/set.cr:
```diff
@@ -308,9 +308,7 @@ struct Set(T)
-  def hash
-    @hash.hash
-  end
+  def_hash @hash
```
Missing doc

straight-shoota commented on src/string.cr:
```diff
@@ -3929,15 +3929,8 @@ class String
-      h = 31 * h + c
-    end
-    h
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/struct.cr:
```diff
@@ -70,15 +70,11 @@ struct Struct
-  # See also: `Object#hash`
-  def hash : Int32
-    hash = 0
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/symbol.cr:
```diff
@@ -15,11 +15,8 @@
-  # See also: `Object#hash`.
-  def hash : Int32
-    to_i
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/time.cr:
```diff
@@ -309,9 +309,7 @@ struct Time
-  def hash
-    @encoded
-  end
+  def_hash @encoded
```
Missing doc

straight-shoota commented on src/tuple.cr:
```diff
@@ -307,15 +307,11 @@ struct Tuple
-  # See also: `Object#hash`.
-  def hash
-    hash = 31 * size
+  def hash(hasher)
```
Missing doc

straight-shoota commented on src/xml/namespace.cr:
```diff
@@ -4,9 +4,7 @@ struct XML::Namespace
-  def hash
-    object_id
-  end
+  def_hash object_id
```
Missing doc

straight-shoota commented on src/xml/node.cr:
```diff
@@ -159,10 +159,7 @@ struct XML::Node
-  def hash
-    object_id
-  end
+  def_hash object_id
```
Missing doc

straight-shoota commented on src/xml/node_set.cr:
```diff
@@ -28,9 +28,7 @@ struct XML::NodeSet
-  def hash
-    object_id
-  end
+  def_hash object_id
```
Missing doc

straight-shoota commented on src/yaml/any.cr:
```diff
@@ -194,10 +194,7 @@ struct YAML::Any
-  def hash
-    raw.hash
-  end
+  def_hash raw
```
Missing doc

straight-shoota commented on src/json/any.cr:
```diff
@@ -261,10 +261,7 @@ struct JSON::Any
-  def hash
-    raw.hash
-  end
+  def_hash raw
```
Missing doc

akzhan commented on spec/std/float_spec.cr:
```diff
@@ -206,11 +206,13 @@ describe "Float" do
     it "does for Float32" do
-      1.2_f32.hash.should_not eq(0)
+      1.2_f32.hash.should eq(1.2_f32.hash)
+      1.2_f32.hash.should_not eq(1.3_f32.hash)
```
this expectation is wrong in common case.

asterite replied:
I'll remove this then. Thanks

akzhan commented on spec/std/float_spec.cr:
```diff
@@ -206,11 +206,13 @@ describe "Float" do
     it "does for Float64" do
-      1.2.hash.should_not eq(0)
+      1.2.hash.should eq(1.2.hash)
+      1.2.hash.should_not eq(1.3.hash)
```
ditto

asterite commented:
I forced pushed a few fixes (maybe next time I'll add more commits so it's simpler to review)

@bcardiff I changed the instance variable name of Hasher to `result` to avoid this confusion
@straight-shoota I added some basic docs, just "See `Object#hash(hasher)`. Maybe these docs should actually be hidden, but it's not a big deal
@akzhan I removed those `hash.should_not eq` specs

I also changed one spec that would randomly failed. This is when a hash has the values, for example, `1` and `1.0` as keys. Their value are equal and their hash value, depending on the initial hasher's seed, could result equal, and so could result equal for `Hash`. This is wrong and should probably be improved. We will probably have to introduce an `eql?` or `equal_for_hash` method, similar to Ruby, where `1 == 1.0` is `true` but `1.eql?(1.0)` is `false`. Another way would be to hardcode this logic in `Hash`, though it gets tricky when these values are nested inside other values. Maybe `def_equals` can also automatically define `eql?`. In any case, this discussion belongs to a different issue.

akzhan commented:
when values are equal, their hashes should be equal too (as documented in Crystal documentation).

This is number normalization that was introduced by #4675 pull request, and will be proposed again later.

RX14 commented:
@asterite `===` and `eql?` are probably the wrong names for what I consider to be internal functions. `equal_for_hash` and `equal_for_case` are better names I think.

asterite commented:
@RX14 @akzhan I don't see how float normalization will solve that particular issue. Care to explain?

akzhan commented:
Number::Normalize returns same hashes for same values. 1.hash, 1.0.hash,
1.to_big_f.hash etc. always be equal with number normalization
пн, 11 сент. 2017 г. в 16:35, Ary Borenszweig <notifications@github.com>:

> @RX14 <https://github.com/rx14> @akzhan <https://github.com/akzhan> I
> don't see how float normalization will solve that particular issue. Care to
> explain?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/crystal-lang/crystal/pull/4946#issuecomment-328530933>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAG3hNPGgpVHChbYtKBW_mlJyFDJDy1wks5shTc2gaJpZM4PSeU3>
> .
>

RX14 commented:
The distinction is between float normalization which transforms a float into an unambiguous bitpattern, and number hash normalization which generates a unique hash from a unique numerical value. I guess thats why LLVM uses canonicalization to refer to the former. We should do the same.

asterite commented:
But hash normalization for numbers will make this problem worse, because `1` and `1.0` will have the same hash value and because `1 == 1.0` a `Hash` can't distinguish them. So this:

```crystal
{1 => 'a', 1.0 => 'b'}
```

will end up being:

```crystal
{1.0 => 'b'}
```

because the second key replaces the first one, they are indistinguishable.

One way for that to work is not to use `==` to compare them, but use `.eql?` or `equals_for_hash`.

RX14 commented:
@asterite What we were suggesting is exactly that behaviour: you cannot insert two "ones" into a hash - regardless of their representation.

asterite commented:
@RX14 Sounds good for now. I don't think it's consistent with union types, but we can delay this "issue" (if it is one) for later.

asterite commented:
Merged!

Feel free to open issues regarding number normalization and using other hashing algorithms, or correct hash values for big numbers.

oprypin commented:
@asterite, see #2818 for why normalization is so important

oprypin commented:
In fact, as I understand it, now numbers of two different types like `{1 => 'a', 1.0 => 'b'}` can **undeterministically** with a very low probability have the same hash causing `{1.0 => 'b'}`

oprypin commented on src/crystal/hasher.cr:
```diff
@@ -0,0 +1,59 @@
+  # TODO: the implementation is naive and should probably use
+  # another algorithm like SipHash or FNV.
+
+  @@seed = uninitialized UInt64
```
`@@seed = Random::System.rand(UInt64::MIN..UInt64::MAX)` with #4894

oprypin commented on src/crystal/hasher.cr:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def float(value)
+    @result *= @result * 31 + value.to_f64.unsafe_as(UInt64)
```
Is this `*=` intended?

oprypin commented:
I really dislike that this got merged in 19 hours.

asterite commented:
@oprypin Don't worry, this is not the final code. This just introduces `Hasher` as a unified way to compute a hash. I think providing a `hash(hasher)` method that objects must implement to append a hash value is the way to go. But do you think there's a different way to do it?

oprypin commented:
The general idea is fine, if not the most elegant. It's more about the small things (please re-check that `*=`!)

akzhan commented:
@asterite 
> @oprypin Don't worry, this is not the final code. This just introduces Hasher as a unified way to compute a hash. I think providing a hash(hasher) method that objects must implement to append a hash value is the way to go. But do you think there's a different way to do it?

I'm sorry, but @funny-falcon code was the same in terms of 'final', but a lot better. Your things are great, but now we should repeat all of its great proposals. Ok.

asterite commented:
@akzhan Maybe, but I don't think so. That PR adds a `hash_normalize` methods to numbers, but I don't think that's correct. That will only be used to implement `hash`, so it should be in `Hasher`.

It also introduces `Number::HashNormalize` type. That belongs in `Hasher` too. `unseeded` is not useful either. It also changes `StringPool`, I don't know why.

Anyway, it had a lot more code than what was needed. And it was so big it received many comments until the author got tired. I think it's better to do this by little steps.

ysbaddaden commented:
BTW: I ported SipHash and made a streaming version:
https://gist.github.com/ysbaddaden/d51c66bc8e07d193ecbb9f7620be8ef5

In terms of performance, computing the hash in one bulk was on par with the official C implementation. Streaming the input adds a ~20% overhead, because the state must be regularly updated.

```crystal
key = uninitialized UInt8[16]
SecureRandom.random_bytes(key.to_slice)

hasher = SipHash64(1, 3).new(key)
hasher.update("some ")
hasher.update("incoming ")
hasher.update("data ")
hasher.final # => UInt64
```

ysbaddaden commented:
BTW again: I published a shard for SipHash and HalfSipHash: https://github.com/ysbaddaden/siphash.cr

RX14 commented:
@ysbaddaden any reason why this isn't a PR to replace the default implementation? I think `SipHash13` should be the default (perhaps only) hasher.

akzhan commented:
Yes, it may be good proposal (but @funny-falcon one is better for LLVM i suppose).

Anyway I don't try to propose any implementation until pushed number normalization.

akzhan commented:
And please leave `result` instead of 'final`, there are no reasons to change the naming.
